test/index.js

```diff
@@ -2055,3 +2055,74 @@ relativeTests2.forEach(function (relativeTest) {
     assert.equal(actual, expected, 'format(' + relativeTest[1] + ') == ' + expected + '\nactual:' + actual);
   });
 });
+
+var fileURLToPathTestCases = [
+  // Lowercase ascii alpha
+  { path: '/foo', fileURL: 'file:///foo' },
+  // Uppercase ascii alpha
+  { path: '/FOO', fileURL: 'file:///FOO' },
+  // dir
+  { path: '/dir/foo', fileURL: 'file:///dir/foo' },
+  // trailing separator
+  { path: '/dir/', fileURL: 'file:///dir/' },
+  // dot
+  { path: '/foo.mjs', fileURL: 'file:///foo.mjs' },
+  // space
+  { path: '/foo bar', fileURL: 'file:///foo%20bar' },
+  // question mark
+  { path: '/foo?bar', fileURL: 'file:///foo%3Fbar' },
+  // number sign
+  { path: '/foo#bar', fileURL: 'file:///foo%23bar' },
+  // ampersand
+  { path: '/foo&bar', fileURL: 'file:///foo&bar' },
+  // equals
+  { path: '/foo=bar', fileURL: 'file:///foo=bar' },
+  // colon
+  { path: '/foo:bar', fileURL: 'file:///foo:bar' },
+  // semicolon
+  { path: '/foo;bar', fileURL: 'file:///foo;bar' },
+  // percent
+  { path: '/foo%bar', fileURL: 'file:///foo%25bar' },
+  // backslash
+  { path: '/foo\\bar', fileURL: 'file:///foo%5Cbar' },
+  // backspace
+  { path: '/foo\bbar', fileURL: 'file:///foo%08bar' },
+  // tab
+  { path: '/foo\tbar', fileURL: 'file:///foo%09bar' },
+  // newline
+  { path: '/foo\nbar', fileURL: 'file:///foo%0Abar' },
+  // carriage return
+  { path: '/foo\rbar', fileURL: 'file:///foo%0Dbar' },
+  // latin1
+  { path: '/fóóbàr', fileURL: 'file:///f%C3%B3%C3%B3b%C3%A0r' },
+  // Euro sign (BMP code point)
+  { path: '/€', fileURL: 'file:///%E2%82%AC' },
+  // Rocket emoji (non-BMP code point)
+  { path: '/🚀', fileURL: 'file:///%F0%9F%9A%80' }
+];
+
+fileURLToPathTestCases.forEach(function (fileURLToPathTestCase) {
+  test('fileURLToPath(' + fileURLToPathTestCase.fileURL + ')', function () {
+    var fromString = url.fileURLToPath(fileURLToPathTestCase.fileURL);
+    assert.strictEqual(fromString, fileURLToPathTestCase.path);
+    var fromURL = url.fileURLToPath(new URL(fileURLToPathTestCase.fileURL));
+    assert.strictEqual(fromURL, fileURLToPathTestCase.path);
```

Comment:
this test case needs to be skipped when `typeof URL !== 'function'`, for older nodes

Reply:
Actually, this is a problem with the implementation as well: https://github.com/mischnic/node-url/blob/5a3f3e445f71d688aa945ebd778034bccf8fb4e2/url.js#L794-L795

Reply:
Indeed it is; we can’t add anything that depends on URL without a full polyfill for it.

Reply:
Using `url.parse` instead might be problematic because that behaves differently to the WHATWG one...

Reply:
I've changed it to use `url.parse` instead now. Only protocol, pathname and hostname are used. So I hope his is fine

Reply:
If it behaves differently then doesn’t that mean we’re not matching node’s url module?

Reply:
I remember reading anecdotally that in general `new URL()` and `url.parse` aren't 100% compatible. But I'm not sure where those differences actually lie. e.g. https://twitter.com/jasnell/status/1417498126282149888

Reply:
right - i've seen comments like that as well, which suggests that using `url.parse` is not actually viable, and that we should be using a full URL polyfill.

Reply:
Should I implement that via https://www.npmjs.com/package/whatwg-url? Should it always prefer a global `URL` if available and only fallback to importing the polyfill otherwise?

Personally I want to use this in the browser, so I'd also like to use the `browser` field in package.json to make sure that the browser's native `URL` class is used to not pull in the additional 70kb-200kb of that polyfill.

Reply:
Unfortunately that particular polyfill doesn't work on old enough engines to be viable. We'd need to locate one that works down to node 0.4 and equivalent browsers.

Reply:
I think https://github.com/zloirock/core-js/blob/master/packages/core-js/web/url.js (which imports https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/web.url.js) should be compatible enough?

Reply:
Great! That implies it's possible. We'd need to build our own implementation tho, based on that one is fine, since that's too large a single package to depend on.

Reply:
So even using `require("core-js-pure/web/url")` would be too large to install/load?

Reply:
Yes, there's a ton of reasons I don't want that particular dependency.